pkg/cmd/cli/cmd/rollback.go

```diff
@@ -38,151 +43,267 @@ a human-readable representation of the updated deployment configuration instead
 
 	return cmd
 }
 
-// newHelper makes a hew helper using real clients.
-func newHelper(oClient client.Interface, kClient kclient.Interface) *helper {
-	return &helper{
-		generateRollback: func(namespace string, config *deployapi.DeploymentConfigRollback) (*deployapi.DeploymentConfig, error) {
-			return oClient.DeploymentConfigs(namespace).Rollback(config)
-		},
-		describe: func(config *deployapi.DeploymentConfig) (string, error) {
-			describer := describe.NewDeploymentConfigDescriberForConfig(oClient, kClient, config)
-			return describer.Describe(config.Namespace, config.Name)
-		},
-		updateConfig: func(namespace string, config *deployapi.DeploymentConfig) (*deployapi.DeploymentConfig, error) {
-			return oClient.DeploymentConfigs(namespace).Update(config)
-		},
-	}
-}
+type RollbackOptions struct {
+	Namespace              string
+	TargetName             string
+	DesiredVersion         int
+	Format                 string
+	Template               string
+	DryRun                 bool
+	IncludeTriggers        bool
+	IncludeStrategy        bool
+	IncludeScalingSettings bool
 
-// helper knows how to perform various rollback related tasks.
-type helper struct {
-	// generateRollback generates a rolled back config from the input config
-	generateRollback func(namespace string, config *deployapi.DeploymentConfigRollback) (*deployapi.DeploymentConfig, error)
-	// describe returns the describer output for config
-	describe func(config *deployapi.DeploymentConfig) (string, error)
-	// updateConfig persists config
-	updateConfig func(namespace string, config *deployapi.DeploymentConfig) (*deployapi.DeploymentConfig, error)
+	out        io.Writer
+	oc         client.Interface
+	kc         kclient.Interface
+	getBuilder func() *resource.Builder
 }
 
-// Generate generates a rolled back DeploymentConfig.
-func (r *helper) Generate(namespace string, config *deployapi.DeploymentConfigRollback) (*deployapi.DeploymentConfig, error) {
-	return r.generateRollback(namespace, config)
-}
+func (o *RollbackOptions) Complete(f *clientcmd.Factory, args []string, out io.Writer) error {
+	// Extract basic flags.
+	if len(args) == 1 {
+		o.TargetName = args[0]
+	}
+	namespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	o.Namespace = namespace
+
+	// Set up client based support.
+	mapper, typer := f.Object()
+	o.getBuilder = func() *resource.Builder {
+		return resource.NewBuilder(mapper, typer, f.ClientMapperForCommand())
+	}
 
-// Describe describes a DeploymentConfig.
-func (r *helper) Describe(config *deployapi.DeploymentConfig, out io.Writer) error {
-	description, err := r.describe(config)
+	oClient, kClient, err := f.Clients()
 	if err != nil {
 		return err
 	}
-	out.Write([]byte(description))
+	o.oc = oClient
+	o.kc = kClient
+
+	o.out = out
+	return nil
+}
+
+func (o *RollbackOptions) Validate() error {
```

Comment:
To clarify, yes, they can't be `nil` when you're running the command itself, but if anyone uses your `RollbackOptions` from another command, they'll expect `Validate` to smack them if they built it wrong.